subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c

```diff
@@ -591,6 +599,16 @@ static int firmware_update_cb(uint16_t obj_inst_id, uint8_t *args, uint16_t args
 static void *firmware_get_buf(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
 			      size_t *data_len)
 {
+	while (k_sem_take(&fw_buf_sem, K_SECONDS(1)) != 0) {
+		if (!k_work_is_pending(&push_data.work)) {
```

Comment:
No, because it does not call this module again, if it fails.

`package_write_cb()` is inside Zephyr's firmware object
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_obj_firmware.c#L253-L271
As you can see, there is 3 exit points before it call the `firmware_block_received_cb()` in our module.

The whole sequence can be checked from the `lwm2m_message_handling.c`
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_message_handling.c#L1049-L1242
1. it calls pre-write callback, which is this `firmware_get_buf()` which reserves the semaphore
2. it calls the post-write callback which is the `firmware_block_received_cb()`
3. that callback may call our `firmware_block_received()`, but only if it did not fail earlier.

So the whole purpose of semaphore is to block the second write of valid blocks. Or just delay, not completely block.

Reply:
Thanks for the detailed explanation!